File: DSA/DrishtiSeth_590011761/Day_13/Question1.py
import re
from collections import Counter

class Solution:
    def mostCommonWord(self, paragraph, banned):
        banned = set(word.lower() for word in banned)

        words = re.findall(r"[a-zA-Z]+", paragraph.lower())

        count = Counter(word for word in words if word not in banned)

        return count.most_common(1)[0][0]


# Example
paragraph = "Bob hit a ball, the hit BALL flew far after it was hit."
banned = ["hit"]

obj = Solution()
print(obj.mostCommonWord(paragraph, banned))

File: DSA/Vaishnavi_590017838/Day_13/Question1.cpp
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    string mostCommonWord(string paragraph, vector<string>& banned) {
        unordered_set<string> bannedWords;
        for(string word : banned)
            bannedWords.insert(word);
        unordered_map<string, int> wordCount;
        string currentWord = "";
        string answer = "";
        int maxCount = 0;
        paragraph += ' ';
        for(char ch : paragraph) {
            if(isalpha(ch)) {
                currentWord += tolower(ch);
            }
            else {
                if(currentWord != "") {
                    if(bannedWords.find(currentWord) == bannedWords.end()) {
                        wordCount[currentWord]++;
                        if(wordCount[currentWord] > maxCount) {
                            maxCount = wordCount[currentWord];
                            answer = currentWord;
                        }
                    }
                    currentWord = "";
                }
            }
        }
        return answer;
    }
};

int main() {
    Solution obj;
    string paragraph = "Bob hit a ball, the hit BALL flew far after it was hit.";
    vector<string> banned = {"hit"};
    cout << obj.mostCommonWord(paragraph, banned);
    return 0;
}
